pytest 7.0.0 changes

As noted by @dtnguyen2 in https://github.com/sherpa/sherpa/pull/1424#issuecomment-1030032628 we now have pytest 7.0.0 available and used by our pip CI runs. Unfortunately we have some questionable pytest use which causes problems - for instance with pytest 6.2.5 wan run the `test_instrument.py` test:

```
%  pytest --version
pytest 6.2.5
% pytest sherpa/tests/test_instrument.py
==================================================== test session starts =====================================================
platform linux -- Python 3.9.7, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/dburke/sherpa/sherpa-main, configfile: pytest.ini
plugins: doctestplus-0.11.2, forked-1.4.0, xdist-2.5.0, cov-3.0.0
collected 16 items

sherpa/tests/test_instrument.py .............X..                                                                       [100%]

=============================================== 15 passed, 1 xpassed in 0.03s ================================================
```

but if we update to `pytest 7.0.0` then they fail

```
% pip install 'pytest > 6.99'
Collecting pytest>6.99
  Downloading pytest-7.0.0-py3-none-any.whl (296 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.5/296.5 KB 6.0 MB/s eta 0:00:00
Requirement already satisfied: pluggy<2.0,>=0.12 in /home/dburke/anaconda/envs/sherpa-main/lib/python3.9/site-packages (from pytest>6.99) (1.0.0)
Requirement already satisfied: packaging in /home/dburke/anaconda/envs/sherpa-main/lib/python3.9/site-packages (from pytest>6.99) (21.3)
Requirement already satisfied: iniconfig in /home/dburke/anaconda/envs/sherpa-main/lib/python3.9/site-packages (from pytest>6.99) (1.1.1)
Requirement already satisfied: py>=1.8.2 in /home/dburke/anaconda/envs/sherpa-main/lib/python3.9/site-packages (from pytest>6.99) (1.11.0)
Requirement already satisfied: attrs>=19.2.0 in /home/dburke/anaconda/envs/sherpa-main/lib/python3.9/site-packages (from pytest>6.99) (21.4.0)
Requirement already satisfied: tomli>=1.0.0 in /home/dburke/anaconda/envs/sherpa-main/lib/python3.9/site-packages (from pytest>6.99) (1.2.3)
Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/dburke/anaconda/envs/sherpa-main/lib/python3.9/site-packages
(from packaging->pytest>6.99) (3.0.6)
Installing collected packages: pytest
  Attempting uninstall: pytest
    Found existing installation: pytest 6.2.5
    Uninstalling pytest-6.2.5:
      Successfully uninstalled pytest-6.2.5
Successfully installed pytest-7.0.0
% pytest sherpa/tests/test_instrument.py
==================================================== test session starts =====================================================
platform linux -- Python 3.9.7, pytest-7.0.0, pluggy-1.0.0
rootdir: /home/dburke/sherpa/sherpa-main, configfile: pytest.ini
plugins: doctestplus-0.11.2, forked-1.4.0, xdist-2.5.0, cov-3.0.0
collected 16 items

sherpa/tests/test_instrument.py .......FFFF..X..                                                                       [100%]

========================================================== FAILURES ==========================================================
___________________________________________________ test_psf1d_kernel_data ___________________________________________________

caplog = <_pytest.logging.LogCaptureFixture object at 0x7f1c6fcab040>

    def test_psf1d_kernel_data(caplog):
        """Access the kernel data: subkernel=True"""

        k = np.asarray([2, 5, 3])
        x = 5 + np.arange(k.size)
        d = Data1D('my-kernel', x, k)

        m = PSFModel(kernel=d)

        dfold = Data1D('fold', np.arange(10), np.zeros(10))

        with caplog.at_level(logging.INFO, logger='sherpa'):
            ans = m.get_kernel(dfold)

        assert len(caplog.records) == 1
        r = caplog.record_tuples[0]
        assert r[0] == 'sherpa.instrument'
        assert r[1] == logging.INFO
        assert r[2] == "PSF frac: 1.0"

        assert isinstance(ans, Data1D)
        assert ans.name == 'kernel'
        # integers, so treat as exact
        assert (ans.x == np.arange(5, 8)).all()
>       assert pytest.approx(k / k.sum(), ans.y)
E       AssertionError: approx() is not supported in a boolean context.
E       Did you mean: `assert a == approx(b)`?

sherpa/tests/test_instrument.py:202: AssertionError
---------------------------------------------------- Captured stdout call ----------------------------------------------------
PSF frac: 1.0
----------------------------------------------------- Captured log call ------------------------------------------------------
INFO     sherpa.instrument:instrument.py:697 PSF frac: 1.0
__________________________________________________ test_psf1d_kernel_model ___________________________________________________
...
```

This should be an easy fix.

--- a/sherpa/tests/test_instrument.py
+++ b/sherpa/tests/test_instrument.py
@@ -1,5 +1,5 @@
 #
-#  Copyright (C) 2020, 2021
+#  Copyright (C) 2020, 2021, 2022
 #  Smithsonian Astrophysical Observatory
 #
 #
@@ -199,7 +199,7 @@ def test_psf1d_kernel_data(caplog):
     assert ans.name == 'kernel'
     # integers, so treat as exact
     assert (ans.x == np.arange(5, 8)).all()
-    assert pytest.approx(k / k.sum(), ans.y)
+    assert ans.y == pytest.approx(k / k.sum())
     assert ans.staterror is None
     assert ans.syserror is None
 
@@ -231,7 +231,7 @@ def test_psf1d_kernel_model(caplog):
 
     # box1D between 1 and 3 inclusive
     y = np.asarray([0, 1, 1, 1, 0, 0, 0, 0, 0, 0])
-    assert pytest.approx(y / y.sum(), ans.y)
+    assert ans.y == pytest.approx(y / y.sum())
 
     assert ans.staterror is None
     assert ans.syserror is None
@@ -275,12 +275,13 @@ def test_psf2d_kernel_data(caplog):
     assert (ans.x0 == x0).all()
     assert (ans.x1 == x1).all()
 
-    assert pytest.approx(k / k.sum(), ans.y)
+    assert ans.y == pytest.approx(k / k.sum())
 
     assert ans.staterror is None
     assert ans.syserror is None
 
 
+@pytest.mark.xfail  # see #1428
 def test_psf2d_kernel_model(caplog):
     """Access the kernel data: no subkernel"""
 
@@ -317,8 +318,13 @@ def test_psf2d_kernel_model(caplog):
     assert (ans.x0 == d0).all()
     assert (ans.x1 == d1).all()
 
+    # This is the old check, converted to what I assume is meant to be
+    # the correct form - see #1428 - but ans.y is all zeros except for
+    # a single 1 whilst k is not (it has 9 True values and the rest
+    # False). I do not have time to identify the truth here.
+    #
     k = (ans.x0 >= -1) & (ans.x0 <= 1) & (ans.x1 >= 10) & (ans.x1 <= 12)
-    assert pytest.approx(k / k.sum(), ans.y)
+    assert ans.y == pytest.approx(k / k.sum())
 
     assert ans.staterror is None
     assert ans.syserror is None

